src/gui/ui/opengl_setup.cpp

```diff
@@ -0,0 +1,80 @@
+#include "opengl_setup.hpp"
+
+#include "../../logging.hpp"
+
+namespace gui {
+
+GLFWwindow*
+setup_opengl(screen_size_t window_width, screen_size_t window_height) {
+    glEnable(GL_MULTISAMPLE);
+
+    // initialize logging
+    glfwWindowHint(GLFW_OPENGL_DEBUG_CONTEXT, true);
+
+    // Initialise GLFW
+    glewExperimental = true; // Needed for core profile
+    if (!glfwInit()) {
+        LOG_CRITICAL(logging::opengl_logger, "Failed to initialize GLFW");
+        getchar();
+        glfwTerminate();
+        return nullptr;
+    }
+
+    glfwWindowHint(GLFW_SAMPLES, 4);               // anti-aliasing of 4
+    glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, 4); // set Major
+    glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 5); // and Minor version
+    glfwWindowHint(
+        GLFW_OPENGL_FORWARD_COMPAT,
+        GL_TRUE
+    ); // To make MacOS happy; should not be needed
+    glfwWindowHint(
+        GLFW_OPENGL_PROFILE,
+        GLFW_OPENGL_CORE_PROFILE
+    ); // somehow turning on core profiling
+
+    // Create window with graphics context
+    GLFWwindow* window = glfwCreateWindow(
+        window_width, window_height, "Dear ImGui GLFW+OpenGL3 example", NULL, NULL
+    );
+    if (window == NULL)
+        return nullptr;
+    glfwMakeContextCurrent(window);
+    glfwSwapInterval(1); // Enable vsync
+
+    glewExperimental = true; // Needed for core profile
+    if (glewInit() != GLEW_OK) {
+        LOG_CRITICAL(logging::opengl_logger, "Failed to initialize GLEW");
+        getchar();
+        glfwTerminate();
+        return nullptr;
```

Comment:
```suggestion
        glfwTerminate();
        return nullptr;
```

Pls use std optional too